web/src/message_view.ts

```diff
@@ -692,7 +692,30 @@ export let show = (raw_terms: NarrowTerm[], show_opts: ShowMessageViewOpts): voi
                 then_select_offset = opts.then_select_offset;
             }
 
-            {
+            if (select_immediately) {
+                // We can skip the initial fetch since we already have a
+                // message we can render and select and sufficient messages
+                // rendered in the view to provide context around the anchor.
+                //
+                // We don't populate message list from locally cached data
+                // if the stream / topic filter needs to be adjusted to
+                // avoid populating messages for the wrong topic.
+                assert(!filter.requires_adjustment_for_moved_with_target);
+                const has_found_newest = msg_list.data.fetch_status.has_found_newest();
+                const has_found_oldest = msg_list.data.fetch_status.has_found_oldest();
+                const has_complete_history = has_found_newest && has_found_oldest;
+                const has_enough_context =
+                    msg_list.num_items() >= message_fetch.consts.narrow_min_num_message_for_context;
+
+                if (!has_complete_history && !has_enough_context) {
+                    message_fetch.fetch_more_if_required_for_current_msg_list(
+                        has_found_oldest,
+                        has_found_newest,
+                        true,
+                        true,
+                    );
+                }
```

Comment:
I think the fallthrough case might be bugged here; the `else` clause logic doesn't run if the above if statement failed to apply.

Reply:
We actually don't want to contact server if we have complete history or enouch context because in that case, we will automatically fetch more messages if required due to this logic in `scroll_finished` block:

```js

    if (message_lists.current.view.should_fetch_older_messages()) {
        // Subtle note: While we've only checked that we're at the
        // very top of the render window (i.e. there may be some more
        // cached messages to render), it's a good idea to fetch
        // additional message history even if we're not actually at
        // the edge of what we already have from the server.
        message_fetch.maybe_load_older_messages({
            msg_list: message_lists.current,
            msg_list_data: message_lists.current.data,
        });
    }

    if (message_lists.current.view.should_fetch_newer_messages()) {
        // See the similar message_viewport.at_rendered_top block.
        message_fetch.maybe_load_newer_messages({
            msg_list: message_lists.current,
        });
    }
```

Reply:
Added a comment here.

Reply:
OK, thanks, the comment is very helpful. I started a #design thread about if we want to do some double-checking fetches.